x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/SecondaryAuthenticator.java

```diff
@@ -0,0 +1,129 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.authc.support;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.ElasticsearchSecurityException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ContextPreservingActionListener;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.xpack.core.security.SecurityContext;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authc.support.SecondaryAuthentication;
+import org.elasticsearch.xpack.core.security.authc.support.UsernamePasswordToken;
+import org.elasticsearch.xpack.security.authc.AuthenticationService;
+
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+/**
+ * Performs "secondary user authentication" (that is, a second user, _not_ second factor authentication).
+ */
+public class SecondaryAuthenticator implements SecondaryAuthentication.Authenticator {
+
+    /**
+     * The term "Authorization" in the header value is to mimic the standard HTTP "Authorization" header
+     */
+    public static final String SECONDARY_AUTH_HEADER_NAME = "es-secondary-authorization";
+
+    private final Logger logger = LogManager.getLogger();
+    private final SecurityContext securityContext;
+    private final AuthenticationService authenticationService;
+
+    public SecondaryAuthenticator(Settings settings, ThreadContext threadContext, AuthenticationService authenticationService) {
+        this(new SecurityContext(settings, threadContext), authenticationService);
+    }
+
+    public SecondaryAuthenticator(SecurityContext securityContext, AuthenticationService authenticationService) {
+        this.securityContext = securityContext;
+        this.authenticationService = authenticationService;
+    }
+
+    /**
+     * @param listener Handler for the {@link SecondaryAuthentication} object.
+     *                 If the secondary authentication credentials do not exist the thread context, the
+     *                 {@link ActionListener#onResponse(Object)} method is called with a {@code null} authentication value.
+     *                 If the secondary authentication credentials are found in the thread context, but fail to be authenticated, then
+     *                 the failure is returned through {@link ActionListener#onFailure(Exception)}.
+     */
+    @Override
+    public void authenticate(String action, TransportRequest request, ActionListener<SecondaryAuthentication> listener) {
+        authenticate(authListener -> authenticationService.authenticate(action, request, false, authListener), listener);
```

Comment:
Would it be worth adding a comment or jdoc to mention that `allowAnonymous` is set to `false` for a reason?